presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalMergeExchange.java

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class LocalMergeExchange
+{
+    private static final int DEFAULT_MAX_BUFFERED_PAGES = 2;
+
+    private final List<ExchangeBuffer> buffers;
+
+    public LocalMergeExchange(int buffersCount)
+    {
+        this(buffersCount, DEFAULT_MAX_BUFFERED_PAGES);
+    }
+
+    public LocalMergeExchange(int buffersCount, int maxBufferedPagesPerBuffer)
+    {
+        checkArgument(buffersCount > 0, "buffersCount must be greater than zero");
+        checkArgument(maxBufferedPagesPerBuffer > 0, "maxBufferedPagesPerBuffer must be greater than zero");
+        buffers = IntStream.range(0, buffersCount)
+                .mapToObj((i) -> new ExchangeBuffer(maxBufferedPagesPerBuffer))
+                .collect(toImmutableList());
+    }
+
+    public List<ExchangeBuffer> getBuffers()
+    {
+        return buffers;
+    }
+
+    public ExchangeBuffer getBuffer(int index)
+    {
+        checkElementIndex(index, buffers.size());
+        return buffers.get(index);
+    }
+
+    public void finishRead()
+    {
+        buffers.forEach(ExchangeBuffer::finishRead);
+    }
+
+    public void close()
+    {
+        finishRead();
+    }
+
+    public static class ExchangeBuffer
+    {
+        private final Object lock = new Object();
+        private final int maxBufferedPages;
+
+        @GuardedBy("lock")
+        private final Queue<Page> pages = new LinkedList<>();
+        @GuardedBy("lock")
+        private SettableFuture<?> readBlocked = SettableFuture.create();
+        @GuardedBy("lock")
+        private SettableFuture<?> writeBlocked = SettableFuture.create();
+        @GuardedBy("lock")
+        private boolean readFinished;
+        @GuardedBy("lock")
+        private boolean writeFinished;
+
+        public ExchangeBuffer(int maxBufferedPages)
+        {
+            checkArgument(maxBufferedPages > 0, "maxBufferedPages must be greater then zero");
+            this.maxBufferedPages = maxBufferedPages;
+            writeBlocked.set(null);
+        }
+
+        public ListenableFuture<?> isWriteBlocked()
+        {
+            synchronized (lock) {
```

Comment:
Although through the presto code `synchronized` method are widely use i find them obscure. By defining method to be synchronized you making your synchronization policy to be public. So it is possible (highly unlikely though) to synchronize on the same object in other bit of code and introduce a deadlock. Kinda paranoid, yes. What do you think, @martint ?